Reorganize realtime_support repository

jacquelinekay opened the pull request:
(note the repository name change)

Moves rttest to a subfolder, create tlsf_c repository (snapshot from http://www.gii.upv.es/tlsf/main/repo), and tlsf_cpp repository with allocator wrapper and example program.

TODO:

tlsf_cpp:
- [ ] reorg allocator example to be a test or gtest using the no new/deletes/mallocs test fixture by @wjwwood
- [x] ~~create test for tlsf_stack_allocator as well as tlsf_heap_allocator~~ I decided not to implement tlsf_stack_allocator, because I don't know how to write a copy constructor for it.
- [x] pass a pool size as a template parameter to allocator example

I'd be open to a different package organization scheme, e.g. make an "allocators" package instead of tlsf_cpp.

dirk-thomas commented:
If huge code blocks are coming from external source I would consider keeping them in a separate repo (similar to gtest_vendor, gmock_vendor).

jacquelinekay commented:
alright, I've pushed the external package to its own repository (ros2/ros2#166)

jacquelinekay commented:
Do you have any idea why this package would fail to find rmw_implementation? http://ci.ros2.org/job/ci_linux/698/console#console-section-5

It builds on my linux desktop.

dirk-thomas commented:
Since the package doesn't depend on `rmw_implementation` it can't `find_package` it when being built in isolation.

I assume you build without isolation locally (which you should)?

jacquelinekay commented:
whoops, yeah, that would make sense.

jacquelinekay commented:
CI:
http://ci.ros2.org/job/ci_linux/710/
http://ci.ros2.org/job/ci_osx/589
http://ci.ros2.org/job/ci_windows_opensplice/778

This package and tlsf are disabled on OSX and Windows (the allocator contains some GNU specific headers).

gerkey commented on tlsf_cpp/example/allocator_example.cpp:
```diff
@@ -0,0 +1,325 @@
+/** Check a demangled stack backtrace of the caller function for the given tokens. */
+bool check_stacktrace(const char ** tokens, size_t num_tokens, size_t max_frames = 8)
+{
+# ifndef WIN32
```
Since `backtrace()` is a GNU extension (at least according to the man page), how about `#ifdef __GNUC__` instead?

gerkey commented on tlsf_cpp/example/allocator_example.cpp:
```diff
@@ -0,0 +1,325 @@
+  void * addrlist[max_frames + 1];
+
+  // retrieve current stack addresses
+  int addrlen = backtrace(addrlist, sizeof(addrlist) / sizeof(void *));
```
It would be a bit more in line with the style guide (https://google.github.io/styleguide/cppguide.html#sizeof) to say `sizeof(addrlist) / sizeof(addrlist[0])`.

gerkey commented on tlsf_cpp/example/allocator_example.cpp:
```diff
@@ -0,0 +1,325 @@
+  free(symbollist);
+  if (!match) {
+    fprintf(stderr, "%s\n", output_buffer);
+  }
```
`free(output_buffer);`

gerkey replied:
Resolved on the following line.

gerkey commented on tlsf_cpp/include/tlsf_cpp/tlsf.hpp:
```diff
@@ -0,0 +1,148 @@
+  tlsf_heap_allocator()
+  : memory_pool(nullptr)
+  {
+    initialize(1024 * 1024);
```
Where does this size (`1024 * 1024`) come from? Should it be exposed or parameterized somehow?

jacquelinekay replied:
Well...

One annoying thing about how the C++ standard library handles allocators is that many data structures that take an allocator, such as `std::vector`, take the allocator type as a template argument, and then default-construct their allocator in their own constructor.

So most instantiations of the allocator will use this default constructor, which is really not awesome.

I previously had the idea to add a template argument to the allocator so that the user has a better path to initializing an allocator, even one that gets default-constructed. I can try implementing that now...

gerkey commented on tlsf_cpp/include/tlsf_cpp/tlsf.hpp:
```diff
@@ -0,0 +1,148 @@
+  // Needed for std::allocator_traits
+  void deallocate(T * ptr, size_t size)
+  {
+    (void)size;
```
Is that to avoid a warning about an unused argument? Can you instead leave out the argument name in the function signature?

jacquelinekay replied:
Unfortunately, the function signature needs to be exactly this way because std::allocator_traits expects an allocator to have a function with this signature, to be used to deallocate memory.

dirk-thomas replied:
You could only have the types in the signature but without a variable name. But in header we usually don't do that since then you can't document the arguments in the (not yet existing) docblock. If the documentation comes e.g. from a parent class it would be an option.

jacquelinekay replied:
Oh, I just reread your comment. Yeah, I can leave out the argument name (that's already being done several places in this file already).

gerkey commented:
I made one minor tweak directly and added a few comments. Otherwise, +1.

Also, wow, that's all kind of intense. I particularly enjoyed the in-process backtrace symbol parser.

jacquelinekay commented:
Yeah, my hope is actually to get rid of the backtracer soon with some pending changes I have to the rmw repositories (although it may still be needed if this test is going to pass with rmw_fastrtps).

Much of that logic is duplicated in the allocator test in test_rclcpp (https://github.com/ros2/system_tests/blob/master/test_rclcpp/test/test_allocator.cpp). In fact, it has lagged behind a bit due to recent changes I made. I'm considering getting rid of the test in the other package and turning this "example" into an actual gtest. I may also move the allocator_example in the rclcpp_examples repository into this one.

jacquelinekay commented:
I updated the backtracer code and moved it into a gtest.

I also noticed a problem fixed in ros2/rclcpp/#187. This pull request is required for the test to pass. I didn't notice it before because I wasn't testing with intraprocess.

I'm removing the allocator test that's in test_rclcpp in ros2/system_tests#91, since the gtest here is strictly stronger than that one and there's a lot of duplicate code between them.

jacquelinekay commented:
I aliased the branch in ros2/rclcpp#187 to a matching branchname and I'm going to re-start CI.

jacquelinekay commented:
http://ci.ros2.org/job/ci_linux/761/
http://ci.ros2.org/job/ci_osx/641/
http://ci.ros2.org/job/ci_windows/826

wjwwood commented:
lgtm, is there anything that needs to be done to get this merged?

jacquelinekay commented:
It needs ros2/rclcpp#187 for the test to pass but otherwise I'm happy with
it

On Thu, Dec 17, 2015 at 2:49 PM, William Woodall <notifications@github.com>
wrote:

> lgtm, is there anything that needs to be done to get this merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/realtime_support/pull/16#issuecomment-165606591>.
>
